Add Clan Moulder

MichaelBrown08 opened the pull request.

msolefonte commented on src/script/vco-ie/main_warhammer/wh2_main_skv_clan_moulder/missions.lua:
```diff
@@ -0,0 +1,146 @@
+					type DESTROY_FACTION;
+
+					faction wh3_main_ksl_the_ice_court;
+                    faction wh3_main_ksl_the_great_orthodoxy;
```
Do not play with papa's OCD

MichaelBrown08 replied:
Will try and be more aware of these!

msolefonte commented on src/script/vco-ie/main_warhammer/wh2_main_skv_clan_moulder/missions.lua:
```diff
@@ -118,8 +118,6 @@ local missions = {
 
 					faction wh_dlc08_nor_wintertooth;
-
-					confederation_valid;
```
Leave it there. I never tested in on Warhammer III, but in old games it could be buggy if another AI confederated the faction, as you could not "destroy it" anymore. Probably not required anymore, but nice to have just as a safeguard.

msolefonte commented on src/script/campaign/mod/vco-listeners.lua:
```diff
@@ -51,6 +51,12 @@ end
 
 -- CHECKS --
 
+local function check_skv_mdr_number_upgrades_purchased(faction_key)
```
Leave the vco there. Even if these are local functions, better to keep them consistent:
`check_vco_skv_mdr_number_upgrades_purchased(faction_key)`

msolefonte commented on src/script/campaign/mod/vco-listeners.lua:
```diff
@@ -155,6 +161,23 @@ end
 local function add_listeners()
 	vco:log("Adding listeners");
 
+	vco:log("- Moulder listeners");
```
Listeners are triggered per race more than per faction, so "Skaven listeners". Also they are sorted alphabetically.

msolefonte commented on src/script/campaign/mod/vco-listeners.lua:
```diff
@@ -155,6 +161,23 @@ end
+            return faction:name() == "wh2_main_skv_clan_moulder" and faction:is_human()
+        end,
+        function(context)
+            vco:log("inside mdr listener execution")
```
Not checking the code/logs yet because I assume it is WIP, but will come back once the release is ready. In any case, first function is a filter and second unit is the the actual callback. We do not want callbacks to hold any code, so just leave a function there. If it has any params, use something like:
```lua
function(context)
	check_vco_brt_bordeleaux_alberic_vow(context:character());
end,
```
else
```lua
check_vco_brt_bordeleaux_alberic_vow,
```

msolefonte commented on src/script/vco-ie/main_warhammer/wh2_main_skv_clan_moulder/missions.lua:
```diff
@@ -0,0 +1,144 @@
+					type DESTROY_FACTION;
+
+					faction wh3_main_ksl_the_ice_court;
+          faction wh3_main_ksl_the_great_orthodoxy;
```
You fixed it wrong. Use tabs instead of spaces for this project. Not sure how to configure VSCode for that. If a problem, simply copy paste the key above as is properly tabulated.

msolefonte commented on src/script/vco-ie/main_warhammer/wh2_main_skv_clan_moulder/missions.lua:
```diff
@@ -0,0 +1,146 @@
+                    faction wh3_main_ksl_the_great_orthodoxy;
+				}
+            
+                objective
```
This whole objective block is tabulated using spaces too.

MichaelBrown08 replied:
VSCode problems. I have no idea how to switch to tabs - I'll look into it.

msolefonte reviewed (commented):
Modify `frontend_faction_leaders_tables` to set `difficulty` as `difficulty` for this faction.

msolefonte commented:
Also, add the markdowns to the `docs` folder. Check the other ones for proper formating.

MichaelBrown08 commented:
I'll add the docs once the placeholders are replaced I think.

msolefonte commented on src/db/vco_victory_types.loc.tsv:
```diff
@@ -102,4 +102,7 @@ victory_types_display_name_vco_victory_type_grn_gri_2_burn_badlands	Route II - B
 victory_types_display_name_vco_victory_type_ogr_gre_1_rich_walk	Route I - Too Rich to Walk	true
 victory_types_display_name_vco_victory_type_ogr_gre_2_overtyrants_reign	Route II - Overtyrant's Reign	true
-victory_types_display_name_vco_victory_type_ogr_gre_3_drakecrush	Route III - Drakecrush	true
+victory_types_display_name_vco_victory_type_ogr_gre_3_drakecrush	Route III - Drakecrusher	true
```
Like 105 is out of scope

MichaelBrown08 commented:
Added a new table in the above commit ☝🏻 "src/db/effect_bonus_value_pooled_resource_factor_junctions_tables_vco.tsv". 

There's no skaven food source from victory conditions so I've added 
* `skaven_food_victory` as a way of getting (resource) `skaven_food`.
*  (effect) `vco_skv_pooled_resource_skaven_food_victory` and it's corresponding link: 
*  `effect_bonus_value_pooled_resource_factor_junctions_tables` from (effect) `vco_skv_pooled_resource_skaven_food_victory` to (resource factor) `skaven_food_victory`.
* And finally added (effect) `vco_skv_pooled_resource_skaven_food_victory` to the (effect bundles) entry for route 2.

It's not displaying anything in the route yet so I need to look into why, but this mirrors the setup for other types of skaven food sources so I _think_ it's right.

MichaelBrown08 commented:
@msolefonte 

What's left to do:

Route 1:
Objective: Unlock all Infantry Augments and Monster Augments (Needs scripting I think)

Route 2:
Reward: Food per turn (I've added 10 food per turn from victory conditions resource factor, just unsure why it's not working?)
Reward: Recruit rank +X for mutants - I assume mutants are the units given from the growth vat? Otherwise if mutants are 'any unit with a flesh lab effect' this needs scripting.

Route 3:
Objective: Can't add Warhost to DESTROY_FACTION? I've tried `wh_main_chs_chaos` and `wh3_dlc20_chs_chaos_qb4` (only non-named chaos faction from wh3 and has flag path `ui\flags\wh_main_chs_chaos`) and both failed?
Reward:  I've tried both `wh2_dlc16_throt_flesh_lab_mutagen_base_income` and `wh2_dlc16_throt_flesh_lab_mutagen_building_income` for mutagen income and neither shows up on the reward. `wh2_dlc16_throt_flesh_lab_growth_vat_mutagen_dummy` shows as "Mutagen By-product +10 (factionwide)" which doesn't look like per-turn mutagen?
Reward: +10% WS or replenishment for mutants (Don't think these effects exist by default, closest I can find is `wh2_dlc16_skv_effect_force_replenishment_rate_moulder` and that's "replenishment for wolf rats, rat ogres, mutant rat ogres, brood horrors, hell pit abominations". Needs scripting?)

I've had a look at unit sets for mutants but unless `wh2_dlc16_throt_flesh_lab_pool` is mutants (I think it's units from the flesh lab?) there doesn't appear to be one.

I think I've done as much as I can here. I'll move onto the orthodoxy.

MichaelBrown08 commented:
Icons should also be checked and updated to appropriate icons - I'll revisit these shortly.

msolefonte commented on docs/immortal-empires/skaven.md:
```diff
@@ -0,0 +1,23 @@
+## Throt the Unclean - Clan Moulder
+
+* **Route I - Yes-yes, prepare for war!**
```
As mentioned on Discord, I do not really like this route name because it sounds like it is a preparation for the other routes. Each route should feel like a unique and independent campaign.

msolefonte replied:
Same with the rest of the flavour regarding this route.

msolefonte commented on src/db/effect_bundles_tables_vco.tsv:
```diff
@@ -29,3 +29,6 @@ vco_victory_payload_ogr_gre_1_rich_walk	Hoardmaster	Greasus' wealth is now so ob
 vco_victory_payload_ogr_gre_2_overtyrants_reign_1	Hell on Wheels	In recognition of Greasus' indisputable dominion over these lands, the Black Dwarfs of Zhar-Naggrund have gifted him with a powerful daemon engine to strap to the back of his cart. In return, they merely ask that he look the other way when the occasional Cathayan caravan goes missing...	battle_movement_character.png	faction	1	true	f ...
 vco_victory_payload_ogr_gre_2_overtyrants_reign_2	Blood-forged Gut-plate	Grimgor's armour was forged in the blood of Dwarf Runepriests, thus granting it potent defensive properties. As it was far too small for Greasus' magnificent bulk, it was melted down and reforged into a gut-plate of terrific proportions.	armour_character.png	faction	1	true	false	true
 vco_victory_payload_ogr_gre_3_drakecrush	Subservient Cathayans	The Cathayans have now accepted their place at the Overtyrant's feet. Caravans journey to his Great Hall laden with gold, gemstones and exotic meats, hoping to barter for safe passage across the Ivory Road. Each season, their profit margins grow smaller.	diplomacy_trespass.png	faction	1	true	false	true
+vco_victory_payload_skv_mdr_1_yes_yes_prepare_for_war_1	War Footing	Throt's mutant armies are ready and the Vermintide awaits. Hell Pit has geared up for war!	growth_juice.png	faction	1	true	false	true
```
Why the `_1`? I can understand it when you get two or more effect bundles for the same route, but I do not think it should be there for the others. At least to keep it coherent with the already existent ones.

msolefonte commented on src/db/effects_tables_vco.tsv:
```diff
@@ -16,3 +16,4 @@ vco_ogr_greasus_speed	battle_movement_character.png	101	battle_movement_characte
 vco_ogr_greasus_charge_bonus	charge_character.png	101	charge_character.png	campaign	true
 vco_ogr_greasus_armour	armour_character.png	101	armour_character.png	campaign	true
 vco_ogr_greasus_ward_save	resistance_ward_save.png	101	resistance_ward_save.png	campaign	true
+vco_skv_pooled_resource_skaven_food_victory	pooled_resource_food.png	101	pooled_resource_food.png	campaign	true
```
W: I do not like the `vco_skv_pooled_resource_skaven_food_victory` becuase it is redudant to have a `pooled_resource` key to include `pooled_resource`. I would change it to `vco_skv_per_turn_food`. It would also be more coherent with the other pooled resource effects: https://github.com/msolefonte/vco/blob/main/src/db/effects_tables_vco.tsv#L15

Note that all these changes about keys can be solved with find and replace direclty in the TSVs.

msolefonte commented on src/db/pooled_resource_factor_junctions_tables_vco.tsv:
```diff
@@ -2,3 +2,4 @@ resource	factor	unique_id	minimum	maximum	specific_faction_set
 #pooled_resource_factor_junctions_tables;2;db/pooled_resource_factor_junctions_tables/vco					
 emp_progress	vco	wh2_dlc13_resource_factor_vco	0	2147483647	
 brt_chivalry	vco	chivalry_vco	-2147483647	2147483647	
+skaven_food	vco	skaven_food_victory	-2147483647	2147483647	
```
W: Again, for the sake of coherence: `skv_food` and `food_vco`/`skv_food_vco`/`skaven_food_vco`.

MichaelBrown08 replied:
These are named as CA name them - I copied their naming pattern. Happy to change them.

MichaelBrown08 replied:
Re: `skaven_food` vs `skv_food` it's already coded as `skaven_food` by CA, despite the rest of the Skaven resources being prefixed with `skv_`.

![image](https://user-images.githubusercontent.com/46346590/191860143-04d9d827-69ba-4fdc-8eaa-14a59ea5057b.png)

MichaelBrown08 replied:
I've updated the rest of the names to be consistent with previous naming patterns however.

msolefonte commented on src/db/vco_effects.loc.tsv:
```diff
@@ -16,3 +16,4 @@ effects_description_vco_ogr_greasus_speed	Speed: %+n for Greasus Goldtooth	true
 effects_description_vco_ogr_greasus_charge_bonus	Charge Bonus: %+n for Greasus Goldtooth	true
 effects_description_vco_ogr_greasus_armour	Armour: %+n for Greasus Goldtooth	true
 effects_description_vco_ogr_greasus_ward_save	Ward save: %+n% for Greasus Goldtooth	true
+effects_description_vco_skv_pooled_resource_skaven_food_victory	Food rewarded for Victory Conditions	true
```
Do not mention "Victory Conditions" in the flavour. It should also be more verbose by including the actual ammount: `Food: %+n per turn`

msolefonte commented on src/script/campaign/mod/vco-listeners.lua:
```diff
@@ -112,6 +112,12 @@ local function check_vco_ogr_goldtooth_gross_income(target_faction)
 	end
 end
 
+local function check_vco_skv_mdr_number_upgrades_purchased(faction_key)
```
I will create an independent ticket for me to fix this one.

msolefonte replied:
Ticket #132 created

msolefonte commented on src/script/campaign/mod/vco-listeners.lua:
```diff
@@ -256,6 +262,22 @@ local function add_listeners()
+		function(context)
+			vco:log("inside mdr listener conditional check")
+			local faction = context:unit():faction()
+			return faction:name() == "wh2_main_skv_clan_moulder" and faction:is_human()
```
S: Now that I think about it, checking if the faction is human before checking the name is more efficient. We should actually modify it in all the listeners. 

Also please add `;` at the end of the lines. I know this is lua and it is not required, but it makes papa happy.

msolefonte reviewed (changes requested):
I left a lot of comments but do not worry about them! I am really impressed by how quick you have been able to get it all, including custom effects and pooled resources. Good stuff! Leave the script part for me and focus on Rasputin now.

msolefonte commented on src/db/effect_bonus_value_pooled_resource_factor_junctions_tables_vco.tsv:
```diff
@@ -0,0 +1,3 @@
+effect	bonus_value_id	resource_factor
```
Already did the merge myself so you do not have to worry about it :D

MichaelBrown08 commented:
@msolefonte Have resolved all comments, noting the `skaven_food` resource CA created: https://github.com/msolefonte/vco/pull/119#discussion_r978129139

Should now be scripting left.

msolefonte commented:
Good stuff. I will take it from now. 

Fuck CA for having inconsistent keys tho

msolefonte commented on docs/immortal-empires/skaven.md:
```diff
@@ -0,0 +1,23 @@
+## Throt the Unclean - Clan Moulder
+
+* **Route I - Yes-yes, prepare for war!**
```
You changed the tables but forgot to update the markdown, for both route name and rewards. Nice changes tho.

msolefonte replied:
Will do myself

msolefonte commented on src/script/vco-ie/main_warhammer/wh2_main_skv_clan_moulder/missions.lua:
```diff
@@ -0,0 +1,150 @@
+					script_key vco_dummy;
+				}
+
+				objective
```
Check if you can override text inside the DESTROY_FACTION objective inside of using a dummy. I think it is possible, if not we should update the wiki. Also add the confederation_valid flag.

msolefonte commented on src/script/vco-ie/main_warhammer/wh2_main_skv_clan_moulder/missions.lua:
```diff
@@ -0,0 +1,150 @@
+				}
+
+				objective
+				{
```
Same as above, check if these can be merged into one.
